README.md
# WSDOT Android App #

Washington State Traveler information provide by The Washington State Department of Transportation.

* Written in Java
* Developed with Android Studio IDE

Features - [View on Google Play](https://play.google.com/store/apps/details?id=gov.wa.wsdot.android.wsdot&hl=en)
--------------------------------
* Traffic Map with real-time alerts and conditions.
* Ferry schedules, alerts and VesselWatch.
* Mountain Pass condition and weather reports.
* WSDOT Social media feeds for the latest news.
* Toll Rates.
* Canadian Border Waits.
* Amtrak Cascades Schedules. 

Dependencies
------------
* [Google Play Services](https://developers.google.com/android/guides/overview)
  * [Analytics](https://developers.google.com/android/reference/com/google/android/gms/analytics/package-summary)
  * [Location](https://developers.google.com/android/reference/com/google/android/gms/location/package-summary)
  * [Maps](https://developers.google.com/android/reference/com/google/android/gms/maps/package-summary)
  * [Ads](https://developers.google.com/android/reference/com/google/android/gms/ads/package-summary)
* [Android Support Library](https://developer.android.com/topic/libraries/support-library/index.html)
  * [Design](https://developer.android.com/topic/libraries/support-library/features.html#design)
  * [Cardview](https://developer.android.com/topic/libraries/support-library/features.html#v7-cardview)
  * [Recyclerview](https://developer.android.com/topic/libraries/support-library/features.html#v7-recyclerview)
  * [v4 Support Libraries](https://developer.android.com/topic/libraries/support-library/features.html#v4)
  * [v7 appcompat library](https://developer.android.com/topic/libraries/support-library/features.html#v7-appcompat)
* [Firebase (Core and Crash Analytics)](https://firebase.google.com/docs/reference/android/packages)
* [Android lifecycle - Architecture Components](https://developer.android.com/topic/libraries/architecture/index.html)
* [Dagger 2](https://google.github.io/dagger/)
* [Constraint Layout](https://developer.android.com/training/constraint-layout/index.html)
* [Tap Target View](https://github.com/KeepSafe/TapTargetView)
* [jsoup](https://github.com/jhy/jsoup)
* [gson](https://github.com/google/gson)
* [viewpagerindicator](http://viewpagerindicator.com/)

Contributing
------------

Find a bug? Got an idea? Send us a pull request or open an issue and we'll take a look.

License
-------

Copyright (c) 2017 Washington State Department of Transportation

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [5.13.4] - 2019-03-28

# Changes
- updates worker manager to 2.0.0

## [5.13.3] - 2019-03-27

# Fixes
- Border wait signs for southbound.

## [5.13.1] - 2019-03-20

# Changes
- My routes update. Now shows a route report with alerts, travel times, and cameras.
- removes adding favorites on route.

## [5.12.2] - 2019-02-25

# Fixes
- crash when changing favorite border waits order

## [5.12.1] - 2019-02-19

# Adds
- Border waits to favorites list

## [5.11.6] - 2019-02-07

# Fixes
- event banner display issues

## [5.11.5] - 2019-02-05

# Fixes
- notification issue

## [5.11.4] - 2019-01-17

# Adds
- performace monitoring lib

# Changes
- updates worker lib

# Fixes
- issue with new ferry schedule date format

## [5.11.3]

# Changes
- Migrates support libs to AndroidX

## [5.11.1] - 2019-01-02

# Changes
- Moves ferry alert bulletin button onto the sailings screen app bar.

## [5.10.1] - 2018-12-17

# Changes
- Migrated to Firebase Analytics.
- Removes unsupported view pager library in favor of native solution.

## [5.9.3] - 2018-11-26

# Fixes
- Display error causing ad banner to cover content.

## [5.9.2] - 2018-11-14

# Changes
- Removes ViewModel processing work done on main thread to workers.
- Refactored various sections to improve performance.

## [5.8.6] - 2018-10-26

# Fixes
- Adds network security config for http camera links on Android 9.

## [5.8.4] - 2018-10-24 

# Fixes
- Crash when loading map on Android 9.

## [5.8.3] - 2018-10-23 

# Added
- ETA and actual departure times to ferries schedule.
- My location button to vessel watch.
- milepost to cameras.

# Fixes
- Vessel Watch opens to correct location for San Juan Island routes.

## [5.8.1] - 2018-09-17

# Added
- Location based sailing selection.

# Changed
- Ferries section navigation hierarchy.
- Sailing screen has been reduced to a spinner on the departures screen. 

# Removed
- Ferries home screen.

## [5.7.0] - 2018-08-15

# Changed
- Removed Static Map API calls in favor of Maps SDK lite mode. New quota on the static maps API prompted this change. The maps SDK for mobile has unlimited use. 

## [5.6.1] - 2018-08-15

# Added
- Analytics for push notification open events
- Everett to the go to location menu

# Changed
- I-405 and SR 167 now display current toll rates. Users can favorite these rates.
- Increased range for ferry terminal cameras, added logic to ensure only ferry cameras are collected.

## [5.5] - unrelased tolling beta

## [5.4.1] - 2018-06-18

## Changed
- Ferries section updated to handle new date format coming in the future. The current format is a .NET style date string returned from the API. The new format that will be added in the future is "yyyy-MM-dd hh:mm a"

## [5.4.0] - 2018-06-07

## Added
- Push notification event tracking

## Fixed
- Express lanes logic no longer expects exactly two status. (I-90 closed)
- Updates analytics event labeling

## [5.3.9] - 2018-05-30

## Fixed
- null Ad identifier after updating to AAPT 2.

## [5.3.8] - 2018-05-29

## Fixed
- Blogger feed. Now uses Blogger API v3.0.

## [5.3.6] - 2018-05-16

## Changed
- Removes leading zero from departure times.

## Fixed
- Crash caused by accessing a null Google Map.

## [5.3.3] - 2018-05-07

## Fixed
- Analytics labeling.
- Null crash in ferries section.

## [5.3.1] - 2018-05-03

## Fixed
- Auto google analytics paths.
- Timing crash with tap target view.

## [5.3.0] - 2018-05-02

## Added
- Adds FCM push notifications.

## [5.2.6] - 2018-04

## Added
- Adds WSF information agent contact info.

## [5.2.4] - 2018-03-12

## Added
- Users can now swipe between cameras in their favorites list.

## [5.2.2] - 2018-03-06

## Fixed
- Crash in DB migration 6 to 7 introduced in last update. 

## [5.2.1] - 2018-02-27

## Added
- New travel times layout. Travel times are no grouped by start and end location, with different routes included in the group.

## [5.1.2] - 2018-02-14

## Fixed
- Crash when renaming saved routes.

## [5.1.1] - 2018-02-12

## Fixed
- Crash in favorites caused by a variable not being setup in time. 

## [5.1.0] - 2018-02-12

## Added
- Past sailings for the current day now display.
- Events section.

## [5.0.9] - 2018-01-23

## Added
- Custom logs for Crashlytics
- Alerts refresh timer on traffic map

## Fixed
- Amtrak Cascades schedule crash
- Vessel Watch - view model access in timer crash
- Error in "Alerts in this area"
- xml error causing My Routes guide to show in settings menu.

## [5.0.6] - 2018-01-18

## Added
- Firebase Crashlytics

## [5.0.5] - 2018-01-17

## Added
- MVVM pattern. (Model, View, View Model)
- Room persistence library.
- Border wait and ferries view model test.
- Better My Routes instructions.

## [5.0.0-4] - unreleased builds

## [4.11.3] - 2017-09-19

## Fixed
- Crash caused by referencing id not in travel charts fragment view.
- Updated US 97 camera url.

## [4.11.2] - 2017-10-18

## Added
- New Amtrak train numbers.

## Changed
- Moved My Routes button to home screen.
- Updated to Gradle 3.0.0

## [4.11.1] - 2017-09-19

## Fixed
- Happening now section Google Analytics

## [4.11.0] - 2017-09-19

## Changed
- Social media feeds moved into "Happening Now" section accessible from the traffic map.
- Map layer settings added to FAB menu for the traffic map

## Removed
- JBLM callout.
- Flickr feed.
- Jobs and Good To Go twitter feeds.

## [4.10.5] - 2017-09-07
## Changed
- Updated app with new WSDOT api key.

## [4.10.4] - 2017-28-07
## Added 
- HERO report contact info.
## Changed
- Added "traffic" advertisement topic.

## [4.10.3] - unreleased

## [4.10.2] - 2017-08-07
## Changed
- Made My Routes tracking service a foreground service. This prevents Android (especially Android O) from reducing location updates or killing the service while routes are recording.

## [4.10.1] - 2017-08-02
## Added

src/gov/wa/wsdot/android/wsdot/Blog.java
/*
 * Copyright (c) 2012 Washington State Department of Transportation
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 *
 */

package gov.wa.wsdot.android.wsdot;

import gov.wa.wsdot.android.wsdot.shared.BlogItem;
import gov.wa.wsdot.android.wsdot.util.AnalyticsUtils;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLConnection;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

import org.json.JSONArray;
import org.json.JSONObject;

import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.actionbarsherlock.app.SherlockListActivity;
import com.actionbarsherlock.view.Menu;
import com.actionbarsherlock.view.MenuItem;

public class Blog extends SherlockListActivity {
	private static final String DEBUG_TAG = "Blog";
	private ArrayList<BlogItem> blogItems = null;
	private BlogItemAdapter adapter;
	DateFormat parseDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSz"); //e.g. 2010-08-20T10:11:00.000-07:00
	DateFormat displayDateFormat = new SimpleDateFormat("MMMM d, yyyy h:mm a");
	private View mLoadingSpinner;
	
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        
        AnalyticsUtils.getInstance(this).trackPageView("/News & Social Media/Blog");
        
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        
        setContentView(R.layout.fragment_list_with_spinner);
        mLoadingSpinner = findViewById(R.id.loading_spinner);
        
        blogItems = new ArrayList<BlogItem>();
        this.adapter = new BlogItemAdapter(this, R.layout.simple_list_item, blogItems);
        setListAdapter(this.adapter);
        new GetBlogItems().execute();
    }

    @Override
	public boolean onCreateOptionsMenu(Menu menu) {
    	getSupportMenuInflater().inflate(R.menu.refresh, menu);
    	
    	return super.onCreateOptionsMenu(menu);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch(item.getItemId()) {
	    case android.R.id.home:
	    	finish();
	    	return true;		
		case R.id.menu_refresh:
			this.adapter.clear();
			blogItems.clear();
			new GetBlogItems().execute();
		}
		
		return super.onOptionsItemSelected(item);
	}    
    
	@Override
	protected void onListItemClick(ListView l, View v, int position, long id) {
		super.onListItemClick(l, v, position, id);

		Bundle b = new Bundle();
		Intent intent = new Intent(this, BlogItemDetails.class);
		b.putString("title", blogItems.get(position).getTitle());
		b.putString("content", blogItems.get(position).getContent());
		b.putString("link", blogItems.get(position).getLink());
		intent.putExtras(b);
		startActivity(intent);
	}

	private class GetBlogItems extends AsyncTask<String, Integer, String> {

		@Override
		protected void onPreExecute() {
			mLoadingSpinner.setVisibility(View.VISIBLE);
		}
		
	    protected void onCancelled() {
	        Toast.makeText(Blog.this, "Cancelled", Toast.LENGTH_SHORT).show();
	    }
		
		@Override
		protected String doInBackground(String... params) {
			try {
				URL url = new URL("http://wsdotblog.blogspot.com/feeds/posts/default?alt=json");
				URLConnection urlConn = url.openConnection();
				BufferedReader in = new BufferedReader(new InputStreamReader(urlConn.getInputStream()));
				String jsonFile = "";
				String line;
				while ((line = in.readLine()) != null)
					jsonFile += line;
				in.close();
				
				JSONObject obj = new JSONObject(jsonFile);
				JSONObject data = obj.getJSONObject("feed");			
				JSONArray entries = data.getJSONArray("entry");
				blogItems = new ArrayList<BlogItem>();
				BlogItem i = null;
				
				for (int j=0; j < entries.length(); j++) {
					if (!this.isCancelled()) {
						JSONObject entry = entries.getJSONObject(j);
						i = new BlogItem();
						i.setTitle(entry.getJSONObject("title").getString("$t"));
						i.setPublished(entry.getJSONObject("published").getString("$t"));
						i.setContent(entry.getJSONObject("content").getString("$t"));
						i.setLink(entry.getJSONArray("link").getJSONObject(4).getString("href"));
						
						blogItems.add(i);
					} else {
						break;
					}
				}

			} catch (Exception e) {
				Log.e(DEBUG_TAG, "Error in network call", e);
			}
			return null;
		}		

		@Override
		protected void onPostExecute(String result) {
			mLoadingSpinner.setVisibility(View.GONE);
			
			if(blogItems != null && blogItems.size() > 0){
                adapter.notifyDataSetChanged();
                for(int i=0;i<blogItems.size();i++)
                adapter.add(blogItems.get(i));
            }
            adapter.notifyDataSetChanged();
		}
	}   
	
	private class BlogItemAdapter extends ArrayAdapter<BlogItem> {
        //private ArrayList<BlogItem> items;
        private Typeface tf = Typeface.createFromAsset(getAssets(), "fonts/Roboto-Regular.ttf");
        private Typeface tfb = Typeface.createFromAsset(getAssets(), "fonts/Roboto-Bold.ttf");

        public BlogItemAdapter(Context context, int textViewResourceId, ArrayList<BlogItem> items) {
	        super(context, textViewResourceId, items);
	        //this.items = items;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
	        if (convertView == null) {
	            convertView = getLayoutInflater().inflate(R.layout.simple_list_item, null);
	        }
	        BlogItem o = getItem(position);
	        if (o != null) {
	            TextView tt = (TextView) convertView.findViewById(R.id.title);
	            tt.setTypeface(tfb);
	            TextView bt = (TextView) convertView.findViewById(R.id.description);
	            bt.setTypeface(tf);
	            if (tt != null) {
	            	tt.setText(o.getTitle());
	            }
	            if(bt != null){
	            	try {
	            		Date date = parseDateFormat.parse(o.getPublished());
	            		bt.setText(displayDateFormat.format(date));
	            	} catch (Exception e) {
	            		Log.e(DEBUG_TAG, "Error parsing date", e);
	            	}
	            }
	        }
	        return convertView;
        }
	}
	
	public static class ViewHolder {
		public TextView tt;
		public TextView bt;
	}	
}
